src/App.tsx

```diff
@@ -1,48 +1,149 @@
+import { Router, Switch, Route } from "react-router-dom";
+import { Toast } from "react-bootstrap";
+import { createBrowserHistory } from "history";
 
 import Header from "./header/Header";
 import IssuesList from "./issues/IssuesList";
 import IssuePage from "./issues/IssuePage";
-import { AuthProvider } from "./auth/AuthProvider";
+import { AuthProvider, useAuth } from "./auth/AuthProvider";
 import Login from "./auth/Login";
 import PrivateRoute from "./auth/PrivateRoute";
 import Dash from "./dash/Dash";
+import Profile from "./profile/Profile";
+import AdminPanel from "./admin/Admin";
+import { ToastMessage, ToastProvider, useToast } from "./shared/ToastContext";
 import { initLibrary } from "./shared/icons";
 import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
 
 initLibrary();
 
+const history = createBrowserHistory();
+
+const protectedRoutes = ['/dash', '/profile', '/admin'];
+
 const App: React.FC = () => {
+  const auth = useAuth();
+
+  React.useEffect(() => {
+    history.listen((loc) => {
+      if (protectedRoutes.includes(loc.pathname)) {
+        auth.check(true);
+      }
+    });
+  }, []);
+
   return (
-    <AuthProvider>
-      <BrowserRouter>
-        <div className="container">
-          <Header />
-          <div>
-            <Switch>
-              <Route exact path="/">
-                HOME CONTENT
-              </Route>
-              <Route path="/login">
-                <Login />
-              </Route>
-              <Route path="/issues/:issue_id">
-                <IssuePage />
-              </Route>
-              <Route path="/issues">
-                <IssuesList />
-              </Route>
-              <PrivateRoute path="/dash">
-                <Dash />
-              </PrivateRoute>
-            </Switch>
-          </div>
+    <Router history={history}>
+      <div className="container">
+        <Header />
+        <div>
+          <Switch>
+            <Route exact path="/">
+              HOME CONTENT
+            </Route>
+            <Route path="/login">
+              <Login />
+            </Route>
+            <Route path="/issues/:issue_id">
+              <IssuePage />
+            </Route>
+            <Route path="/issues">
+              <IssuesList />
+            </Route>
+            <PrivateRoute path="/dash">
+              <Dash />
+            </PrivateRoute>
+            <PrivateRoute path="/profile">
+              <Profile />
+            </PrivateRoute>
+            <PrivateRoute admin={true} path="/admin">
+              <AdminPanel />
+            </PrivateRoute>
+          </Switch>
         </div>
-      </BrowserRouter>
-    </AuthProvider>
+      </div>
+    </Router>
+  );
+};
+
+const ToastContainer = () => {
+  const toast = useToast();
+
+  return (
+    <div aria-live="polite" aria-atomic="true" style={{
+      position: "fixed",
+      right: 0,
+      bottom: 0,
+    }}>
+      {toast.toasts.map((msg, i) => {
```

Comment:
I respect this algorithm you've built here, but instead of doing what I believe is a linear search, you could just use a map.

Reply:
Because React state depends on comparing objects, using a map wouldn't work, as React would detect the updated state as being the same and wouldn't rerender. To get around this, the DS used needs to be copied, which is an O(n) operation regardless.

I should probably abstract away the heuristics into ToastContext though, it's pretty ugly.

Reply:
0e0707b3